Added improvements in resolving the best route

saledjenic opened the pull request:
This PR adds improvements in resolving the best route for multi tx, should fix https://github.com/status-im/status-desktop/issues/15762

Some new params are also exposed in this PR for convenience:
- `MaxFeesPerGas                   *hexutil.Big   // Max fees per gas (determined by client via GasFeeMode, in ETH WEI)`
- `TxFee   *hexutil.Big // fee for the transaction (includes tx fee only, doesn't include approval fees, l1 fees, l1 approval fees, token fees or bonders fees, in ETH WEI)`
- `ApprovalFee   *hexutil.Big // Total fee for the approval transaction (includes approval tx fees only, doesn't include approval l1 fees, in ETH WEI)`
- `TxTotalFee *hexutil.Big // Total fee for the transaction (includes tx fees, approval fees, l1 fees, l1 approval fees, in ETH WEI)`

This is the issue we had:
<img width="591" alt="_a1" src="https://github.com/user-attachments/assets/06498ea4-90e6-49ab-a514-2316abce7c4b">


And with these changes it looks like this:
<img width="592" alt="_a2" src="https://github.com/user-attachments/assets/15c450bc-729e-4d12-a52f-5fe35859b083">

saledjenic committed:
fix_: filtering out routes which do not match the amount in

saledjenic committed:
fix_: finding the best route logic updated

status-im-auto commented:
### Jenkins Builds
<details>
<summary>Click to see older builds (20)</summary>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | a8dc10ab | [#1](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5569/1/) | 2024-07-24 08:59:02 | ~2 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5569/1//consoleText) |
| :heavy_check_mark: | a8dc10ab | [#1](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5569/1/) | 2024-07-24 09:01:03 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240724-085639-a8dc10-pr5569.zip) |
| :heavy_check_mark: | a8dc10ab | [#1](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5569/1/) | 2024-07-24 09:01:53 | ~5 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240724-085639-a8dc10-pr5569.zip) |
| :heavy_check_mark: | a8dc10ab | [#1](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5569/1/) | 2024-07-24 09:02:38 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240724-085639-a8dc10-pr5569.aar) |
| :heavy_multiplication_x: | a8dc10ab | [#1](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/1/) | 2024-07-24 09:38:06 | ~41 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/1//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 560214ac | [#2](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5569/2/) | 2024-07-24 10:27:41 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240724-102410-560214-pr5569.zip) |
| :heavy_check_mark: | 560214ac | [#2](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5569/2/) | 2024-07-24 10:29:41 | ~5 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240724-102410-560214-pr5569.zip) |
| :heavy_check_mark: | 560214ac | [#2](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5569/2/) | 2024-07-24 10:30:51 | ~6 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5569/2//consoleText) |
| :heavy_check_mark: | 560214ac | [#2](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5569/2/) | 2024-07-24 10:33:09 | ~8 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240724-102410-560214-pr5569.aar) |
| :heavy_multiplication_x: | 560214ac | [#2](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/2/) | 2024-07-24 10:52:02 | ~27 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/2//consoleText) |
| :heavy_multiplication_x: | 560214ac | [#3](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/3/) | 2024-07-24 11:43:04 | ~38 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/3//consoleText) |
| :heavy_multiplication_x: | 560214ac | [#4](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/4/) | 2024-07-24 12:52:57 | ~38 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/4//consoleText) |
| :heavy_multiplication_x: | 560214ac | [#5](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/5/) | 2024-07-24 13:34:28 | ~38 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/5//consoleText) |
| :heavy_multiplication_x: | 560214ac | [#6](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/6/) | 2024-07-24 14:30:22 | ~38 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/6//consoleText) |
| :heavy_multiplication_x: | 560214ac | [#7](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/7/) | 2024-07-24 19:07:18 | ~39 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/7//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 1b402049 | [#3](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5569/3/) | 2024-07-24 21:36:46 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240724-213433-1b4020-pr5569.aar) |
| :heavy_check_mark: | 1b402049 | [#3](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5569/3/) | 2024-07-24 21:36:47 | ~2 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5569/3//consoleText) |
| :heavy_check_mark: | 1b402049 | [#3](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5569/3/) | 2024-07-24 21:37:04 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240724-213437-1b4020-pr5569.zip) |
| :heavy_check_mark: | 1b402049 | [#3](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5569/3/) | 2024-07-24 21:38:10 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240724-213433-1b4020-pr5569.zip) |
| :heavy_multiplication_x: | 1b402049 | [#8](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/8/) | 2024-07-24 22:19:19 | ~44 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/8//consoleText) |
</details>

| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :heavy_check_mark: | 99e35820 | [#4](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5569/4/) | 2024-07-25 07:13:49 | ~2 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5569/4//consoleText) |
| :heavy_check_mark: | 99e35820 | [#4](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5569/4/) | 2024-07-25 07:13:55 | ~2 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240725-071138-99e358-pr5569.aar) |
| :heavy_check_mark: | 99e35820 | [#4](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5569/4/) | 2024-07-25 07:15:19 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240725-071138-99e358-pr5569.zip) |
| :heavy_check_mark: | 99e35820 | [#4](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5569/4/) | 2024-07-25 07:15:43 | ~4 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240725-071138-99e358-pr5569.zip) |
| :heavy_multiplication_x: | 99e35820 | [#9](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/9/) | 2024-07-25 07:50:47 | ~39 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/9//consoleText) |
| :heavy_multiplication_x: | 99e35820 | [#10](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/10/) | 2024-07-25 09:39:04 | ~36 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/10//consoleText) |
| :heavy_multiplication_x: | 99e35820 | [#11](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/11/) | 2024-07-25 11:18:17 | ~38 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/11//consoleText) |
| | | | | | | |
| :heavy_check_mark: | 966fe184 | [#5](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5569/5/) | 2024-07-25 10:46:56 | ~1 min | `tests-rpc` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests-rpc/job/PR-5569/5//consoleText) |
| :heavy_check_mark: | 966fe184 | [#5](https://ci.status.im/job/status-go/job/prs/job/linux/job/PR-5569/5/) | 2024-07-25 10:47:51 | ~2 min | `linux` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-desktop-240725-104535-966fe1-pr5569.zip) |
| :heavy_check_mark: | 966fe184 | [#5](https://ci.status.im/job/status-go/job/prs/job/ios/job/PR-5569/5/) | 2024-07-25 10:49:11 | ~3 min | `ios` | [:package:`zip`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-ios-240725-104535-966fe1-pr5569.zip) |
| :heavy_check_mark: | 966fe184 | [#5](https://ci.status.im/job/status-go/job/prs/job/android/job/PR-5569/5/) | 2024-07-25 10:51:26 | ~5 min | `android` | [:package:`aar`](https://status-im-prs.ams3.cdn.digitaloceanspaces.com/status-go-android-240725-104535-966fe1-pr5569.aar) |
| :heavy_multiplication_x: | 966fe184 | [#12](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/12/) | 2024-07-25 11:57:51 | ~39 min | `tests` | [:page_facing_up:`log`](https://ci.status.im/job/status-go/job/prs/job/tests/job/PR-5569/12//consoleText) |

IvanBelyakoff commented on services/wallet/router/fees.go:
```diff
@@ -54,27 +54,23 @@ type SuggestedFeesGwei struct {
 }
 
-func (s *SuggestedFees) feeFor(mode GasFeeMode) *big.Int {
-	if !s.EIP1559Enabled {
```
May I ask you why it is removed?

saledjenic replied:
Because of implementation simplifications, look at all changes done in that file.
Lines 140-142.

Cuteivist reviewed (approved):
LGTM

IvanBelyakoff reviewed (approved):
LGTM. Could you pleas answer my inline question?

J-Son89 commented:
can you please add a link in the description to the issue this is solving 👍

saledjenic commented:
@J-Son89 added. :)

alaibe commented on services/wallet/router/fees.go:
```diff
@@ -141,9 +137,9 @@ func (f *FeeManager) SuggestedFees(ctx context.Context, chainID uint64) (*Sugges
-				High:   (*hexutil.Big)(big.NewInt(0)),
+				Low:    (*hexutil.Big)(gasPrice),
+				Medium: (*hexutil.Big)(gasPrice),
+				High:   (*hexutil.Big)(gasPrice),
```
that used to be there to detect non eip1559 chain but there is no such chain anymore right?

Shall we return the error?

saledjenic replied:
Everything is the same, I added `func (m *MaxFeesLevels) feeFor(mode GasFeeMode) *big.Int`, that way we're getting appropriate MaxFee based on the passed `GasFeeMode`, then I updated `func (s *SuggestedFees) feeFor(mode GasFeeMode) *big.Int` which previously had `if !s.EIP1559Enabled` condition, under which it was returning `GasPrice`, but now in the case of `EIP1559Enabled` we're returning default `MaxFee` of `MaxFeesLevels ` which is set to `GasPrice`, so basically the effect is the same, just written differently. We can revert this part and nothing will be changed.

saledjenic committed:
fix_: "to" chains being used in sending bridge tx via hop are more explicit

Using `ChainIDTo` depicts better an intention which chain should be used.

saledjenic committed:
chore_: checking for required balance improved

An error contains now more details, for which token on which chain there is not enough balance for
executing a tx. Also check for required balance now calculates in fees for all tx that possibly can be
sent from the same chain.
